File: frontend/src/types/etf.ts
export type EtfBasic = {
  symbol: string;
  code: string;
  exchange: string;
  name: string;
  theme: string;
  full_name: string | null;
  index_code: string | null;
  index_name: string | null;
  manager: string | null;
  list_date: string | null;
  latest_price: number | null;
  pct_chg: number | null;
  volume: number | null;
  amount: number | null;
  status: string | null;
  provider: string;
  updated_at: string | null;
};

export type EtfDailyBar = {
  symbol: string;
  date: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number | null;
  pre_close: number | null;
  change: number | null;
  pct_chg: number | null;
  volume: number | null;
  amount: number | null;
  factor: number;
  provider: string;
  updated_at: string | null;
};

export type DataStatus = {
  provider: string;
  dataset: string;
  last_success_at: string | null;
  last_trade_date: string | null;
  rows: number;
  status: "ok" | "stale" | "empty" | "error";
  message: string;
};

export type RefreshResult = {
  ok: boolean;
  message: string;
  provider: string;
  selected_rows: number;
  daily_rows: number;
  factor_rows: number;
  factor_latest_date: string | null;
  refreshed_at: string | null;
  failures: Array<Record<string, string>>;
};

export type FactorDefinition = {
  name: string;
  label: string;
  description: string;
  lookback_days: string;
  direction: "higher_better" | "lower_better";
  category: "return" | "risk" | "liquidity" | "trend";
  format: "percent" | "amount" | "ratio";
  interpretation: string;
  limitation: string;
};

export type FactorHistoryPoint = {
  date: string;
  factor_value: number;
  rank: number;
  percentile: number;
};

export type FactorScore = {
  symbol: string;
  name: string;
  theme: string;
  date: string;
  factor_name: string;
  factor_value: number;
  rank: number;
  percentile: number;
  lookback_days: number;
  provider: string;
  updated_at: string | null;
};

export type SignalComponent = {
  factor_name: string;
  label: string;
  category: string;
  direction: "higher_better" | "lower_better";
  value_format: "percent" | "amount" | "ratio";
  factor_value: number;
  rank: number;
  percentile: number;
  weight: number;
  contribution: number;
  interpretation: string;
};

export type SignalExplanation = {
  positive_drivers: string[];
  risk_notes: string[];
  validation_notes: string[];
  data_notes: string[];
};

export type MarketRegimeSnapshot = {
  label: "risk_on" | "neutral" | "risk_off" | string;
  label_zh: string;
  date: string | null;
  target_exposure: number;
  trend_score: number | null;
  volatility_annualized: number | null;
  drawdown: number | null;
  source: string;
  data_notes: string[];
};

export type ResearchSignal = {
  symbol: string;
  name: string;
  theme: string;
  date: string;
  research_score: number;
  priority: string;
  components: SignalComponent[];
  explanation: SignalExplanation;
  market_regime: MarketRegimeSnapshot | null;
  target_exposure: number | null;
  provider: string;
  updated_at: string | null;
};

export type FactorDistribution = {
  count: number;
  missing_count: number;
  min: number | null;
  p25: number | null;
  median: number | null;
  p75: number | null;
  max: number | null;
  mean: number | null;
};

export type FactorForwardReturn = {
  horizon: number;
  sample_count: number;
  top_mean: number | null;
  bottom_mean: number | null;
  spread: number | null;
  data_notes: string[];
};

export type FactorStability = {
  lookback_dates: number;
  average_rank_change: number | null;
  label: string;
  data_notes: string[];
};

export type FactorDiagnostics = {
  factor_name: string;
  date: string | null;
  definition: FactorDefinition;
  distribution: FactorDistribution;
  top: FactorScore[];
  bottom: FactorScore[];
  forward_return: FactorForwardReturn;
  stability: FactorStability;
  data_notes: string[];
};

export type FactorICPoint = {
  date: string;
  ic_value: number;
};

export type FactorICStats = {
  factor_name: string;
  period_start: string | null;
  period_end: string | null;
  num_periods: number;
  rank_ic_mean: number | null;
  rank_ic_std: number | null;
  icir: number | null;
  ic_pos_ratio: number | null;
  ic_t_stat: number | null;
  ic_series: FactorICPoint[];
  ic_by_horizon: Record<string, number | null>;
};

export type FactorQuantilePoint = {
  date: string;
  returns: Record<string, number | null>;
};

export type FactorQuantileReturns = {
  factor_name: string;
  horizon_days: number;
  num_quantiles: number;
  quantile_returns: Record<string, number | null>;
  spread: number | null;
  is_monotonic: boolean;
  quantile_series: FactorQuantilePoint[];
};

export type RedundantFactorPair = {
  factor_a: string;
  factor_b: string;
  correlation: number;
};

export type FactorCorrelationMatrix = {
  date: string | null;
  factor_names: string[];
  matrix: Array<Array<number | null>>;
  redundant_pairs: RedundantFactorPair[];
};

export type FactorTurnoverStats = {
  factor_name: string;
  rank_autocorr: Record<string, number | null>;
  avg_turnover: number | null;
  migration_matrix: Array<Array<number | null>>;
};

export type FactorEvaluationReport = {
  factor_name: string;
  generated_at: string;
  period_start: string | null;
  period_end: string | null;
  num_assets_avg: number;
  ic: FactorICStats;
  quantile_returns: Record<string, FactorQuantileReturns>;
  correlations: FactorCorrelationMatrix | null;
  turnover: FactorTurnoverStats | null;
  warnings: string[];
  overall_verdict: "可用" | "待观察" | "不推荐单独使用";
};

export type FactorClusterGroup = {
  group: string;
  factors: string[];
};

export type FactorPoolEvaluationReport = {
  generated_at: string;
  individual_reports: FactorEvaluationReport[];
  correlation_matrix: FactorCorrelationMatrix;
  cluster_groups: FactorClusterGroup[];
  pool_health: "健康" | "冗余" | "因子不足";
  recommendations: string[];
};

export type BacktestConfig = {
  start: string | null;
  end: string | null;
  top_n: number;
  rebalance: string;
  cost_bps: number;
  benchmark: string;
};

export type BacktestMetrics = {
  cumulative_return: number | null;
  annualized_return: number | null;
  annualized_volatility: number | null;
  max_drawdown: number | null;
  sharpe_like: number | null;
  win_rate_daily: number | null;
  average_turnover: number | null;
  rebalance_count: number;
  final_equity: number | null;
};

export type BacktestEquityPoint = {
  date: string;
  equity: number;
  daily_return: number | null;
};

export type BacktestHolding = {
  symbol: string;
  name: string;
  theme: string;
  research_score: number;
  weight: number;
};

export type BacktestHoldingSnapshot = {
  rebalance_date: string;
  effective_date: string | null;
  holdings: BacktestHolding[];
  turnover: number;
  cost: number;
  regime: MarketRegimeSnapshot | null;
  target_exposure: number;
  realized_exposure: number;
  cash_weight: number;
  risk_notes: string[];
};

export type BacktestBenchmark = {
  key: string;
  label: string;
  symbol: string | null;
  metrics: BacktestMetrics;
  equity_curve: BacktestEquityPoint[];
};

export type BacktestDataNote = {
  severity: "info" | "warning" | "error";
  message: string;
};

export type BacktestRiskSummary = {
  average_target_exposure: number | null;
  average_realized_exposure: number | null;
  average_cash_weight: number | null;
  risk_off_rebalance_count: number;
  constrained_rebalance_count: number;
  data_notes: string[];
};

export type BacktestResult = {
  config: BacktestConfig;
  metrics: BacktestMetrics;
  equity_curve: BacktestEquityPoint[];
  benchmarks: BacktestBenchmark[];
  holdings: BacktestHoldingSnapshot[];
  risk_summary: BacktestRiskSummary | null;
  data_notes: BacktestDataNote[];
};

export type WalkForwardConfig = {
  train_months: number;
  test_months: number;
  step_months: number;
  top_n: number;
  cost_bps: number;
};

export type WalkForwardWindow = {
  train_start: string;
  train_end: string;
  test_start: string;
  test_end: string;
  selected_factors: string[];
  factor_weights: Record<string, number>;
  train_icir: Record<string, number | null>;
  test_metrics: BacktestMetrics;
  test_equity_curve: BacktestEquityPoint[];
};

export type WalkForwardResult = {
  config: WalkForwardConfig;
  windows: WalkForwardWindow[];
  oos_equity_curve: BacktestEquityPoint[];
  oos_metrics: BacktestMetrics;
  is_metrics: BacktestMetrics;
  factor_stability: number;
  weight_stability: Record<string, number | null>;
  overfit_warning: string;
  data_notes: string[];
  generated_at: string | null;
};

export type SignalSnapshotResult = {
  ok: boolean;
  date: string | null;
  rows: number;
  message: string;
};

export type SignalPerformance = {
  snapshot_date: string;
  horizon_days: number;
  top_mean_return: number;
  benchmark_return: number;
  excess_return: number;
  hit_rate: number;
  score_return_corr: number | null;
  top_symbols: string[];
  data_notes: string[];
};

export type MonitoringCheck = {
  key: string;
  label: string;
  status: "ok" | "watch" | "alert" | "missing";
  value: string;
  detail: string;
};

export type MonitoringReport = {
  generated_at: string;
  market_regime: MarketRegimeSnapshot | null;
  checks: MonitoringCheck[];
  alerts: string[];
  data_notes: string[];
};

File: backend/app/models/etf.py
from __future__ import annotations

from datetime import date as dt_date
from datetime import datetime

from pydantic import BaseModel, Field


class EtfBasic(BaseModel):
    symbol: str
    code: str
    exchange: str
    name: str
    theme: str = "其他"
    full_name: str | None = None
    index_code: str | None = None
    index_name: str | None = None
    manager: str | None = None
    list_date: dt_date | None = None
    latest_price: float | None = None
    iopv: float | None = None
    premium_discount_rate: float | None = None
    pct_chg: float | None = None
    change: float | None = None
    open: float | None = None
    high: float | None = None
    low: float | None = None
    pre_close: float | None = None
    amplitude: float | None = None
    volume: float | None = None
    amount: float | None = None
    turnover_rate: float | None = None
    volume_ratio: float | None = None
    latest_share: float | None = None
    circulating_market_value: float | None = None
    total_market_value: float | None = None
    spot_date: dt_date | None = None
    quote_updated_at: datetime | None = None
    status: str | None = None
    provider: str = "unknown"
    source_endpoint: str | None = None
    updated_at: datetime | None = None


class EtfDailyBar(BaseModel):
    symbol: str
    date: dt_date
    open: float | None = None
    high: float | None = None
    low: float | None = None
    close: float | None = None
    pre_close: float | None = None
    change: float | None = None
    pct_chg: float | None = None
    volume: float | None = None
    amount: float | None = None
    amplitude: float | None = None
    turnover_rate: float | None = None
    factor: float = 1.0
    provider: str = "unknown"
    source_endpoint: str | None = None
    updated_at: datetime | None = None


class DataStatus(BaseModel):
    provider: str
    dataset: str
    last_success_at: datetime | None = None
    last_trade_date: dt_date | None = None
    rows: int = 0
    status: str = Field(pattern="^(ok|stale|empty|error)$")
    message: str


class DataSourceAudit(BaseModel):
    generated_at: datetime
    ok: bool
    status: str = Field(pattern="^(ok|warning|error)$")
    provider: str = "unknown"
    manifest_status: str | None = None
    collected_at: datetime | None = None
    spot_source: str | None = None
    selected_rows: int = 0
    basic_rows: int = 0
    daily_rows: int = 0
    latest_trade_date: dt_date | None = None
    symbols_total: int = 0
    symbols_with_daily: int = 0
    daily_missing_symbols: list[str] = Field(default_factory=list)
    source_endpoints: list[str] = Field(default_factory=list)
    cached_sources: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)


class RefreshResult(BaseModel):
    ok: bool
    message: str
    provider: str = "akshare"
    selected_rows: int = 0
    daily_rows: int = 0
    factor_rows: int = 0
    factor_latest_date: dt_date | None = None
    refreshed_at: datetime | None = None
    failures: list[dict[str, str]] = Field(default_factory=list)


class RotationScoreModel(BaseModel):
    symbol: str
    name: str
    theme: str
    etf_type: str
    date: dt_date
    total_score: float
    boom_score: float
    momentum_score: float
    valuation_score: float
    structure_score: float
    liquidity_score: float
    risk_score: float
    boom_status: str
    boom_source: str = "data"
    valuation_percentile: float
    state: str
    action: str
    returns: dict[str, float | None] = Field(default_factory=dict)
    risk_notes: list[str] = Field(default_factory=list)
    drivers: list[str] = Field(default_factory=list)
    input_notes: str = ""


class RotationReportModel(BaseModel):
    radar_date: dt_date | None = None
    rankings: list[RotationScoreModel] = Field(default_factory=list)
    pools: dict[str, list[str]] = Field(default_factory=dict)
    data_notes: list[str] = Field(default_factory=list)


class FactorScore(BaseModel):
    symbol: str
    name: str
    theme: str
    date: dt_date
    factor_name: str
    factor_value: float
    rank: int
    percentile: float
    lookback_days: int = 0
    provider: str = "local"
    updated_at: datetime | None = None


class FactorHistoryPoint(BaseModel):
    date: dt_date
    factor_value: float
    rank: int
    percentile: float


class FactorDistribution(BaseModel):
    count: int = 0
    missing_count: int = 0
    min: float | None = None
    p25: float | None = None
    median: float | None = None
    p75: float | None = None
    max: float | None = None
    mean: float | None = None


class FactorForwardReturn(BaseModel):
    horizon: int
    sample_count: int = 0
    top_mean: float | None = None
    bottom_mean: float | None = None
    spread: float | None = None
    data_notes: list[str] = Field(default_factory=list)


class FactorStability(BaseModel):
    lookback_dates: int = 0
    average_rank_change: float | None = None
    label: str = "样本不足"
    data_notes: list[str] = Field(default_factory=list)


class FactorDiagnostics(BaseModel):
    factor_name: str
    date: dt_date | None = None
    definition: dict[str, object]
    distribution: FactorDistribution
    top: list[FactorScore] = Field(default_factory=list)
    bottom: list[FactorScore] = Field(default_factory=list)
    forward_return: FactorForwardReturn
    stability: FactorStability
    data_notes: list[str] = Field(default_factory=list)


class SignalComponent(BaseModel):
    factor_name: str
    label: str
    category: str
    direction: str
    value_format: str
    factor_value: float
    rank: int
    percentile: float
    weight: float
    contribution: float
    interpretation: str


class SignalExplanation(BaseModel):
    positive_drivers: list[str] = Field(default_factory=list)
    risk_notes: list[str] = Field(default_factory=list)
    validation_notes: list[str] = Field(default_factory=list)
    data_notes: list[str] = Field(default_factory=list)


class MarketRegimeSnapshot(BaseModel):
    label: str = "neutral"
    label_zh: str = "中性"
    date: dt_date | None = None
    target_exposure: float = 0.75
    trend_score: float | None = None
    volatility_annualized: float | None = None
    drawdown: float | None = None
    source: str = "universe_equal_weight"
    data_notes: list[str] = Field(default_factory=list)


class ResearchSignal(BaseModel):
    symbol: str
    name: str
    theme: str
    date: dt_date
    research_score: float
    priority: str
    components: list[SignalComponent] = Field(default_factory=list)
    explanation: SignalExplanation
    market_regime: MarketRegimeSnapshot | None = None
    target_exposure: float | None = None
    provider: str = "local"
    updated_at: datetime | None = None


class BacktestConfig(BaseModel):
    start: dt_date | None = None
    end: dt_date | None = None
    top_n: int = 10
    rebalance: str = "monthly"
    cost_bps: float = 5.0
    benchmark: str = "510300.SH"


class BacktestMetrics(BaseModel):
    cumulative_return: float | None = None
    annualized_return: float | None = None
    annualized_volatility: float | None = None
    max_drawdown: float | None = None
    sharpe_like: float | None = None
    win_rate_daily: float | None = None
    average_turnover: float | None = None
    rebalance_count: int = 0
    final_equity: float | None = None


class BacktestEquityPoint(BaseModel):
    date: dt_date
    equity: float
    daily_return: float | None = None


class BacktestHolding(BaseModel):
    symbol: str
    name: str
    theme: str
    research_score: float
    weight: float


class BacktestHoldingSnapshot(BaseModel):
    rebalance_date: dt_date
    effective_date: dt_date | None = None
    holdings: list[BacktestHolding] = Field(default_factory=list)
    turnover: float = 0.0
    cost: float = 0.0
    regime: MarketRegimeSnapshot | None = None
    target_exposure: float = 1.0
    realized_exposure: float = 1.0
    cash_weight: float = 0.0
    risk_notes: list[str] = Field(default_factory=list)


class BacktestBenchmark(BaseModel):
    key: str
    label: str
    symbol: str | None = None
    metrics: BacktestMetrics
    equity_curve: list[BacktestEquityPoint] = Field(default_factory=list)


class BacktestDataNote(BaseModel):
    severity: str = "info"
    message: str


class BacktestRiskSummary(BaseModel):
    average_target_exposure: float | None = None
    average_realized_exposure: float | None = None
    average_cash_weight: float | None = None
    risk_off_rebalance_count: int = 0
    constrained_rebalance_count: int = 0
    data_notes: list[str] = Field(default_factory=list)


class BacktestResult(BaseModel):
    config: BacktestConfig
    metrics: BacktestMetrics
    equity_curve: list[BacktestEquityPoint] = Field(default_factory=list)
    benchmarks: list[BacktestBenchmark] = Field(default_factory=list)
    holdings: list[BacktestHoldingSnapshot] = Field(default_factory=list)
    risk_summary: BacktestRiskSummary | None = None
    data_notes: list[BacktestDataNote] = Field(default_factory=list)


class SignalSnapshotResult(BaseModel):
    ok: bool
    date: str | None = None
    rows: int = 0
    message: str = ""


class SignalPerformanceModel(BaseModel):
    snapshot_date: dt_date
    horizon_days: int
    top_mean_return: float = 0.0
    benchmark_return: float = 0.0
    excess_return: float = 0.0
    hit_rate: float = 0.0
    score_return_corr: float | None = None
    top_symbols: list[str] = Field(default_factory=list)
    data_notes: list[str] = Field(default_factory=list)


class MonitoringCheck(BaseModel):
    key: str
    label: str
    status: str = Field(pattern="^(ok|watch|alert|missing)$")
    value: str = ""
    detail: str = ""


class MonitoringReport(BaseModel):
    generated_at: datetime
    market_regime: MarketRegimeSnapshot | None = None
    checks: list[MonitoringCheck] = Field(default_factory=list)
    alerts: list[str] = Field(default_factory=list)
    data_notes: list[str] = Field(default_factory=list)
